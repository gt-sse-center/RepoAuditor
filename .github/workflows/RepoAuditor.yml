name: RepoAuditor

on: 
  workflow_dispatch:
  schedule:
    - cron: '0 0 * * *' # Every day at midnight UTC

jobs:
  audit-repo:
    runs-on: ubuntu-latest
    steps:
      - uses: mtanneau/RepoAuditor@action-v0.2.1
        with:
          github_pat: ${{ secrets.REPO_AUDITOR_PAT }}
