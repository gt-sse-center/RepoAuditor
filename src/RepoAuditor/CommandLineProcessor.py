# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""Contains the CommandLineProcessor object"""

from dataclasses import dataclass, field
from typing import Any, Protocol

from dbrownell_Common.Streams.DoneManager import DoneManager  # type: ignore[import-untyped]

from .ExecuteModules import Execute, Module, ModuleInfo


# ----------------------------------------------------------------------
@dataclass(frozen=True)
class CommandLineProcessor:
    """\
    Object that makes it easy to process command line arguments and invoke ExecuteModules.Execute
    in a testable way.
    """

    # ----------------------------------------------------------------------
    # |
    # |  Public Types
    # |
    # ----------------------------------------------------------------------
    class GetDynamicArgsFunc(Protocol):
        def __call__(
            self,
            dynamic_arg_definitions: dict[
                str, Any
            ],  # actual type is dict[str, TyperEx.TypeDefinitionItemType]
        ) -> dict[str, Any]: ...

    # ----------------------------------------------------------------------
    # |
    # |  Public Data
    # |
    # ----------------------------------------------------------------------
    module_infos: list[ModuleInfo]
    warnings_as_error_module_names: set[str]
    ignore_warnings_module_names: set[str]
    single_threaded: bool = field(kw_only=True)

    # ----------------------------------------------------------------------
    # |
    # |  Public Methods
    # |
    # ----------------------------------------------------------------------
    @classmethod
    def Create(
        cls,
        get_dynamic_args_func: "CommandLineProcessor.GetDynamicArgsFunc",
        modules: list[Module],
        includes: list[str],
        excludes: list[str],
        warnings_as_error_module_names: set[str],
        ignore_warnings_module_names: set[str],
        *,
        all_warnings_as_error: bool = False,
        ignore_all_warnings: bool = False,
        single_threaded: bool = False,
        argument_separator: str = "-",
    ) -> "CommandLineProcessor":
        # Convert modules into a lookup map
        module_map: dict[str, Module] = {}

        for module in modules:
            if argument_separator in module.name:
                raise Exception(f"The module name '{module.name}' contains '{argument_separator}'.")

            prev_module = module_map.get(module.name, None)
            if prev_module is not None:
                raise Exception(f"The module '{module.name}' has already been defined.")

            module_map[module.name] = module

        del modules

        # Process includes
        included_modules: set[str] = set()
        included_requirements: dict[str, set[str]] = {}

        for include in includes:
            parts = include.split(argument_separator)

            this_module = module_map.get(parts[0], None)
            if this_module is None:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 1:
                included_modules.add(parts[0])
            else:
                included_requirements.setdefault(this_module.name, set()).add(
                    argument_separator.join(parts[1:])
                )

        del includes

        for module_name, module in list(module_map.items()):
            if module.requires_explicit_include and module_name not in included_modules:
                module_map.pop(module_name)
                continue

        del included_modules

        # Process excludes
        excluded_requirements: dict[str, set[str]] = {}

        for exclude in excludes:
            parts = exclude.split(argument_separator)

            this_module = module_map.get(parts[0], None)
            if this_module is None:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 1:
                module_map.pop(parts[0])
            else:
                excluded_requirements.setdefault(this_module.name, set()).add(
                    argument_separator.join(parts[1:])
                )

        del excludes

        # Add / Remove the requirements for each module
        for module_name, module in list(module_map.items()):
            module.ProcessRequirements(
                included_requirements.get(module_name, set()),
                excluded_requirements.get(module_name, set()),
            )

            if module.GetNumRequirements() == 0:
                module_map.pop(module_name)

        del included_requirements
        del excluded_requirements

        if all_warnings_as_error:
            warnings_as_error_module_names = {module.name for module in module_map.values()}
        if ignore_all_warnings:
            ignore_warnings_module_names = {module.name for module in module_map.values()}

        # Create the type definitions
        dynamic_arg_definitions: dict[str, Any] = {}

        for module in module_map.values():
            # Module-level args
            for key, value in module.GetDynamicArgDefinitions().items():
                dynamic_arg_definitions[f"{module.name}{argument_separator}{key}"] = value

            # Requirement-level args
            for query in module.queries:
                for requirement in query.requirements:
                    for key, value in requirement.GetDynamicArgDefinitions().items():
                        dynamic_arg_definitions[
                            f"{module.name}{argument_separator}{requirement.name}{argument_separator}{key}"
                        ] = value

        # Process the dynamic info
        dynamic_args: dict[str, dict[str, Any]] = {}

        for key, value in get_dynamic_args_func(dynamic_arg_definitions).items():
            parts = key.split(argument_separator)
            assert len(parts) >= 2

            if parts[0] not in module_map:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 2:
                dynamic_args.setdefault(parts[0], {})[parts[1]] = value
            else:
                dynamic_args.setdefault(parts[0], {}).setdefault(None, {}).setdefault(  # type: ignore
                    parts[1],
                    {},
                )[argument_separator.join(parts[2:])] = value

        module_infos: list[ModuleInfo] = []

        for module_name, module in module_map.items():
            module_args = dynamic_args.get(module_name, {})
            requirement_args = module_args.pop(None, None)  # type: ignore

            module_infos.append(ModuleInfo(module, module_args, requirement_args))

        return cls(
            module_infos,
            warnings_as_error_module_names,
            ignore_warnings_module_names,
            single_threaded=single_threaded,
        )

    # ----------------------------------------------------------------------
    def __call__(
        self,
        dm: DoneManager,
    ) -> list[list[Module.EvaluateInfo]]:
        return Execute(dm, self.module_infos)
