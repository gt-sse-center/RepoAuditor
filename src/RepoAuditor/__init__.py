# pylint: disable=missing-module-docstring,invalid-name

APP_NAME = "RepoAuditor"

# Note that this value will be overwritten by calls to `python ../../Build.py update_version` based
# on changes observed in the git repository. The default value below will be used until the value
# here is explicitly updated by the Continuous Integration system.
__version__ = "0.1.24"
