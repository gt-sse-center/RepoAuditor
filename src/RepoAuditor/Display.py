# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""Contains functionality to display the results of executing modules."""

import textwrap
from typing import Optional

from dbrownell_Common import TextwrapEx  # type: ignore[import-untyped]
from dbrownell_Common.Streams.DoneManager import DoneManager  # type: ignore[import-untyped]
from rich import print as rich_print
from rich.console import Group
from rich.panel import Panel

from RepoAuditor.Module import EvaluateResult, Module


def GetInternalPanelContent(
    dm: DoneManager,
    results: list[Module.EvaluateInfo],
    display_resolution: bool,  # noqa: FBT001
    display_rationale: bool,  # noqa: FBT001
    num_success: int,
    num_warning: int,
    num_error: int,
    num_does_not_apply: int,
    num_requirements: int,
) -> list[Panel | str]:
    """Get the content for each result to be displayed inside the internal panel."""
    internal_content: list[Panel | str] = []

    module = results[0].module

    if module.description:
        internal_content.append(f"{module.description.strip()}\n")

    internal_content.append(
        textwrap.dedent(
            f"""\
                Successful:     {num_success} ({(num_success / num_requirements):.02%})
                Warnings:       {num_warning} ({(num_warning / num_requirements):.02%})
                Errors:         {num_error} ({(num_error / num_requirements):.02%})
                Skipped:        {num_does_not_apply} ({(num_does_not_apply / num_requirements):.02%})
                """,
        ),
    )

    for result in results:
        if not dm.is_verbose and result.result in [
            EvaluateResult.Success,
            EvaluateResult.DoesNotApply,
        ]:
            continue

        if result.result == EvaluateResult.Success:
            border_color = "green"
        elif result.result == EvaluateResult.Warning:
            border_color = "yellow"
        elif result.result == EvaluateResult.Error:
            border_color = "red"
        elif result.result == EvaluateResult.DoesNotApply:
            border_color = ""
        else:
            raise RuntimeError(result.result)  # pragma: no cover

        content: list[str] = []

        if result.requirement.description:
            content.append(f"{result.requirement.description.strip()}\n")

        if result.context:
            content.append(f"{result.result.name.upper()}: {result.context.strip()}\n")

        if display_resolution and result.resolution:
            content.append(
                textwrap.dedent(
                    """\
                        Resolution
                        ----------
                          {}
                        """,
                ).format(TextwrapEx.Indent(result.resolution.strip(), 2, skip_first_line=True)),
            )

        if display_rationale and result.rationale:
            content.append(
                textwrap.dedent(
                    """\
                        Rationale
                        ---------
                          {}
                        """,
                ).format(TextwrapEx.Indent(result.rationale.strip(), 2, skip_first_line=True)),
            )

        internal_content.append(
            Panel(
                Group(*content),
                title=f"[{result.result.name}] {result.requirement.name}",
                title_align="left",
                border_style=border_color,
                padding=(1, 1, 0, 1) if content else 0,
            ),
        )

    return internal_content


# ----------------------------------------------------------------------
def DisplayResults(
    dm: DoneManager,
    all_results: list[list[Module.EvaluateInfo]],
    *,
    display_resolution: bool,
    display_rationale: bool,
    panel_width: Optional[int] = None,
) -> None:
    """Display the results of executing the modules."""
    for results in all_results:
        assert results
        module = results[0].module

        num_success = 0
        num_warning = 0
        num_error = 0
        num_does_not_apply = 0

        for result in results:
            if result.result == EvaluateResult.Success:
                num_success += 1
            elif result.result == EvaluateResult.Warning:
                num_warning += 1
            elif result.result == EvaluateResult.Error:
                num_error += 1
            elif result.result == EvaluateResult.DoesNotApply:
                num_does_not_apply += 1
            else:
                raise RuntimeError(result.result)  # pragma: no cover

        num_requirements = module.GetNumRequirements()
        assert num_requirements

        internal_content: list[Panel | str] = GetInternalPanelContent(
            dm,
            results,
            display_resolution,
            display_rationale,
            num_success=num_success,
            num_warning=num_warning,
            num_error=num_error,
            num_does_not_apply=num_does_not_apply,
            num_requirements=num_requirements,
        )

        rich_print(
            Panel(
                Group(*internal_content),
                padding=1 if internal_content else 0,
                title=results[0].module.name,
                width=panel_width,
            ),
        )
