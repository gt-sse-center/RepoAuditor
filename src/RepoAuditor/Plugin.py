# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""Contains the interface that Plugins must implement."""

import pluggy

from RepoAuditor import APP_NAME
from RepoAuditor.Module import Module


# ----------------------------------------------------------------------
@pluggy.HookspecMarker(APP_NAME)
def GetModule() -> Module:
    """Return a Module."""
    raise NotImplementedError("hookspec")  # pragma: no cover # noqa: EM101
