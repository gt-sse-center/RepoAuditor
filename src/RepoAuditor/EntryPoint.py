# ----------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# ----------------------------------------------------------------------
"""Audits a repository against a set of requirements."""

import sys
import textwrap
import traceback

from typing import Annotated

import pluggy
import typer

from click.exceptions import UsageError
from dbrownell_Common.Streams.DoneManager import DoneManager, Flags as DoneManagerFlags  # type: ignore [import-untyped]
from dbrownell_Common import TextwrapEx  # type: ignore [import-untyped]
from dbrownell_Common import TyperEx  # type: ignore [import-untyped]
from typer.core import TyperGroup  # type: ignore [import-untyped]

from RepoAuditor import APP_NAME, __version__
from RepoAuditor.CommandLineProcessor import CommandLineProcessor, Module
from RepoAuditor.ExecuteModules import DisplayResults
from RepoAuditor import Plugin


# ----------------------------------------------------------------------
ARGUMENT_SEPARATOR = "-"


# ----------------------------------------------------------------------
class NaturalOrderGrouper(TyperGroup):
    # ----------------------------------------------------------------------
    def list_commands(self, *args, **kwargs):
        return self.commands.keys()


# ----------------------------------------------------------------------
app = typer.Typer(
    cls=NaturalOrderGrouper,
    help=__doc__,
    no_args_is_help=True,
    pretty_exceptions_show_locals=False,
    pretty_exceptions_enable=False,
)


# ----------------------------------------------------------------------
def _GetModules() -> list[Module]:
    plugin_manager = pluggy.PluginManager(APP_NAME)

    plugin_manager.add_hookspecs(Plugin)
    plugin_manager.load_setuptools_entrypoints(APP_NAME)

    return plugin_manager.hook.GetModule()


_all_modules = _GetModules()
del _GetModules


# ----------------------------------------------------------------------
def _HelpEpilog() -> str:
    content: list[str] = []

    # ----------------------------------------------------------------------
    def TypeInfoToString(
        arg_name: str,
        type_info: TyperEx.TypeDefinitionItemType,
    ) -> str:
        if isinstance(type_info, TyperEx.TypeDefinitionItem):
            python_type = type_info.python_type
            parameter_info = type_info.parameter_info
        elif isinstance(type_info, tuple):
            python_type = type_info[0]
            parameter_info = type_info[1]
        else:
            python_type = type_info
            parameter_info = None

        return "    {arg_name:<50} {type_description:<7} {help}".format(
            arg_name=arg_name,
            type_description=python_type.__name__,
            help="" if parameter_info is None else parameter_info.help,
        )

    # ----------------------------------------------------------------------

    for module in _all_modules:
        # Get the module arguments
        module_arguments: list[str] = []

        for arg_name, type_info in module.GetDynamicArgDefinitions().items():
            module_arguments.append(
                TypeInfoToString(
                    f"--{module.name}{ARGUMENT_SEPARATOR}{arg_name}",
                    type_info,
                )
            )

        final_module_arguments = TextwrapEx.Indent(
            "\n".join(module_arguments),
            4,
            skip_first_line=True,
        )

        # Get the requirement arguments
        requirement_arguments: list[str] = []

        for query in module.queries:
            for requirement in query.requirements:
                for key, value in requirement.GetDynamicArgDefinitions().items():
                    requirement_arguments.append(
                        TypeInfoToString(
                            f"--{module.name}{ARGUMENT_SEPARATOR}{requirement.name}{ARGUMENT_SEPARATOR}{key}",
                            value,
                        ),
                    )

        final_requirement_arguments = TextwrapEx.Indent(
            "\n".join(requirement_arguments),
            4,
            skip_first_line=True,
        )

        content.append(
            textwrap.dedent(
                f"""\
                {"-" * len(module.name)}
                {module.name}
                {"-" * len(module.name)}
                {module.description}

                Command Line Arguments:
                    {final_module_arguments}

                Optional Arguments:
                    {final_requirement_arguments}
                """,
            ),
        )

    return (
        textwrap.dedent(
            """\
        Module Information
        ==================

        {}

        Command Line Examples
        =====================
        repo_auditor --include GitHub --GitHub-url https://github.com/gt-sse-center/RepoAuditor --GitHub-pat <PAT or path to file containing PAT>

        """,
        )
        .format(
            "\n".join(content),
        )
        .replace("\n", "\n\n")
    )


# ----------------------------------------------------------------------
def _VersionCallback(value: bool) -> None:
    if value:
        sys.stdout.write(f"RepoAuditor v{__version__}\n")
        raise typer.Exit()


# ----------------------------------------------------------------------
@app.command(
    "EntryPoint",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
    help=__doc__,
    epilog=_HelpEpilog(),
    no_args_is_help=False,
)
def EntryPoint(
    ctx: typer.Context,
    version: Annotated[
        bool,
        typer.Option(
            "--version",
            help="Display the version of this tool and exit.",
            callback=_VersionCallback,
            is_eager=True,
        ),
    ] = False,
    includes: Annotated[
        list[str],
        typer.Option(
            "--include",
            help=f"Module or requirement names to explicitly include in the execution; like other command line arguments, requirement names must include the module name as a prefix (e.g. 'ModuleName{ARGUMENT_SEPARATOR}RequirementName'). This value can be provided multiple times.",
        ),
    ] = [],
    excludes: Annotated[
        list[str],
        typer.Option(
            "--exclude",
            help=f"Module or requirement names to exclude from execution; like other command line arguments, requirement names must include the module name as a prefix (e.g. 'ModuleName{ARGUMENT_SEPARATOR}RequirementName'). This value can be provided multiple times.",
        ),
    ] = [],
    warnings_as_error: Annotated[
        list[str],
        typer.Option(
            "--warnings-as-error",
            help="Name of a module whose warnings should be treated as errors. This value can be provided multiple times.",
        ),
    ] = [],
    ignore_warnings: Annotated[
        list[str],
        typer.Option(
            "--ignore-warnings",
            help="Name of a module whose warnings should be ignored. This value can be provided multiple times.",
        ),
    ] = [],
    all_warnings_as_error: Annotated[
        bool, typer.Option("--all-warnings-as-error", help="Treat all warnings as errors.")
    ] = False,
    ignore_all_warnings: Annotated[
        bool, typer.Option("--ignore-all-warnings", help="Ignore all warnings.")
    ] = False,
    single_threaded: Annotated[
        bool,
        typer.Option("--single-threaded", help="Do not use multiple threads when evaluating requirements."),
    ] = False,
    no_resolution: Annotated[
        bool,
        typer.Option(
            "--no-resolution",
            help="Do not display resolution information for requirements that are not successful.",
        ),
    ] = False,
    no_rationale: Annotated[
        bool,
        typer.Option(
            "--no-rationale",
            help="Do not display rationale information for requirements that are not successful.",
        ),
    ] = False,
    verbose: Annotated[
        bool,
        typer.Option(
            "--verbose",
            help="Write verbose information to the terminal.",
        ),
    ] = False,
    debug: Annotated[
        bool,
        typer.Option(
            "--debug",
            help="Write debug information to the terminal.",
        ),
    ] = False,
) -> None:
    with DoneManager.CreateCommandLine(
        sys.stdout,
        flags=DoneManagerFlags.Create(verbose=verbose, debug=debug),
    ) as dm:
        try:
            executor = CommandLineProcessor.Create(
                lambda dynamic_arg_definitions: TyperEx.ProcessDynamicArgs(ctx, dynamic_arg_definitions),
                _all_modules,
                includes,
                excludes,
                set(warnings_as_error),
                set(ignore_warnings),
                all_warnings_as_error=all_warnings_as_error,
                ignore_all_warnings=ignore_all_warnings,
                single_threaded=single_threaded,
                argument_separator=ARGUMENT_SEPARATOR,
            )

        except Exception as ex:
            if dm.is_debug:
                raise

            raise UsageError(str(ex)) from ex

        try:
            all_results = executor(dm)

            dm.WriteLine("\n\n")

            DisplayResults(
                dm,
                all_results,
                display_resolution=not no_resolution,
                display_rationale=not no_rationale,
            )
        except Exception as ex:
            if dm.is_debug:
                error = traceback.format_exc()
            else:
                error = str(ex)

            dm.WriteError(error)


# ----------------------------------------------------------------------
# ----------------------------------------------------------------------
# ----------------------------------------------------------------------
if __name__ == "__main__":
    app()  # pragma: no cover
