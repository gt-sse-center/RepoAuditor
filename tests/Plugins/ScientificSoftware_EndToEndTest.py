# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""End-to-end tests for the ScientificSoftware plugin."""

import pytest
from dbrownell_Common.TestHelpers.StreamTestHelpers import (
    InitializeStreamCapabilities,
)
from typer.testing import CliRunner
from utilities import ScrubDurationGithuburlAndSpaces

from RepoAuditor.EntryPoint import app

# ----------------------------------------------------------------------
pytest.fixture(InitializeStreamCapabilities(), scope="session", autouse=True)


# ----------------------------------------------------------------------
class TestScientificSoftware:
    """End-to-end tests for files in a GitHub repository which implements scientific software."""

    # ----------------------------------------------------------------------
    def test_NoCitation(self, args, snapshot):
        """Test if no CITATION file exists in the repository."""
        result = CliRunner().invoke(
            app,
            args,
        )

        assert result.exit_code == -1, result.output
        assert ScrubDurationGithuburlAndSpaces(result.stdout) == snapshot
