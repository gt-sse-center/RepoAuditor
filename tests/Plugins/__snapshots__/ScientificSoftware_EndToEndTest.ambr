# serializer version: 1
# name: TestScientificSoftware.test_NoCitation
  '''
  Processing 1 module...
    Processing 'ScientificSoftware' (1 of 1)...
  
  
  
  
    DONE! ([31;1m-1[0m, <scrubbed duration>)
  DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ ScientificSoftware â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
  â”‚                                                                                                                      â”‚
  â”‚ Validates existence of repository files for Scientific Software.                                                     â”‚
  â”‚                                                                                                                      â”‚
  â”‚                                                                                                                      â”‚
  â”‚ [37mSuccessful:     0 (0.00%)[0m                                                                                            â”‚
  â”‚ [37mWarnings:       0 (0.00%)[0m                                                                                            â”‚
  â”‚ [37mErrors:         1 (100.00%)[0m                                                                                          â”‚
  â”‚ [37mSkipped:        0 (0.00%)[0m                                                                                            â”‚
  â”‚                                                                                                                      â”‚
  â”‚ [31mâ•­â”€[0m[31m [Error] Citation [0m[31mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m[31mâ”€â•®[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Validates that CITATION file exists.                                                                             [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ERROR: No CITATION file found.                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Resolution                                                                                                       [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ----------                                                                                                       [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   1) Create a CITATION/CITATIONS file in in the root directory.                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   2) It can have any of the following extensions:                                                                [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m      -  e.g. `CITATION`                                                                                          [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m      - CITATION.cff                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m      - CITATION.md                                                                                               [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m      - CITATION.bib                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Rationale                                                                                                        [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ---------                                                                                                        [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   A citation file helps users correctly cite your software.                                                      [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   For more details, please see                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repos [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m itory/about-citation-files                                                                                       [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[0m â”‚
  â”‚ [37mâ•­â”€[0m[37m Metrics [0m[37mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m[37mâ”€â•®[0m â”‚
  â”‚ [37mâ”‚[0m                                                                                                                  [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m                                                                                                                  [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m [37mSuccessful:     0 (0.00%)[0m                                                                                        [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m [37mWarnings:       0 (0.00%)[0m                                                                                        [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m [37mErrors:         1 (100.00%)[0m                                                                                      [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m [37mSkipped:        0 (0.00%)[0m                                                                                        [37mâ”‚[0m â”‚
  â”‚ [37mâ”‚[0m                                                                                                                  [37mâ”‚[0m â”‚
  â”‚ [37mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[0m â”‚
  â”‚                                                                                                                      â”‚
  â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
  
  Results: DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  '''
# ---
